src/index.js

```diff
@@ -63,327 +65,244 @@ class PubsubBaseProtocol extends EventEmitter {
   }
 
   /**
-   * On successful connection event.
+   * Registrar notifies a closing connection with pubsub protocol.
    * @private
-   * @param {String} protocol connection protocol
-   * @param {Connection} conn connection to the peer
+   * @param {PeerInfo} peerInfo peer info
+   * @param {Error} err error for connection end
    */
-  _onConnection (protocol, conn) {
-    conn.getPeerInfo((err, peerInfo) => {
-      if (err) {
-        this.log.err('Failed to identify incomming conn', err)
-        return pull(empty(), conn)
-      }
-
-      const idB58Str = peerInfo.id.toB58String()
-      const peer = this._addPeer(new Peer(peerInfo))
+  _onPeerDisconnected (peerInfo, err) {
+    const idB58Str = peerInfo.id.toB58String()
+    const peer = this.peers.get(idB58Str)
 
-      this._processConnection(idB58Str, conn, peer)
-    })
+    this.log('connection ended', idB58Str, err ? err.message : '')
+    this._removePeer(peer)
   }
 
   /**
-   * Overriding the implementation of _processConnection should keep the connection and is
-   * responsible for processing each RPC message received by other peers.
-   * @abstract
-   * @param {string} idB58Str peer id string in base58
-   * @param {Connection} conn connection
+   * Add a new connected peer to the peers map.
+   * @private
    * @param {PeerInfo} peer peer info
-   * @returns {undefined}
-   *
+   * @returns {PeerInfo}
    */
-  _processConnection (idB58Str, conn, peer) {
-    throw errcode('_processConnection must be implemented by the subclass', 'ERR_NOT_IMPLEMENTED')
+  _addPeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    let existing = this.peers.get(id)
+    if (!existing) {
+      this.log('new peer', id)
+      this.peers.set(id, peer)
+      existing = peer
+
+      peer.once('close', () => this._removePeer(peer))
+    }
+
+    return existing
   }
 
   /**
-   * On connection end event.
+   * Remove a peer from the peers map.
    * @private
-   * @param {string} idB58Str peer id string in base58
-   * @param {PeerInfo} peer peer info
-   * @param {Error} err error for connection end
+   * @param {Peer} peer peer state
+   * @returns {PeerInfo}
    */
-  _onConnectionEnd (idB58Str, peer, err) {
-    // socket hang up, means the one side canceled
-    if (err && err.message !== 'socket hang up') {
-      this.log.err(err)
+  _removePeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    this.log('delete peer', id)
+    this.peers.delete(id)
+    return peer
+  }
+
+  /**
+   * Validates the given message. The signature will be checked for authenticity.
+   * @param {rpc.RPC.Message} message
+   * @returns {Promise<Boolean>}
+   */
+  async validate (message) { // eslint-disable-line require-await
+    // If strict signing is on and we have no signature, abort
+    if (this.strictSigning && !message.signature) {
+      this.log('Signing required and no signature was present, dropping message:', message)
+      return Promise.resolve(false)
```

Comment:
Just return false, `Promise.resolve` is unnecessary. 
```suggestion
      return false
```